File: backend/app/analysis/pipeline.py
"""End-to-end signal analysis pipeline."""

from __future__ import annotations

from uuid import UUID, uuid4
from datetime import datetime

import structlog

from app.analysis.confidence import calculate_confidence
from app.analysis.fact_extraction import extract_facts
from app.analysis.sentiment import classify_sentiment
from app.analysis.themes import identify_themes
from app.llm.models import LLMRequest
from app.llm.prompts import build_summary_prompt
from app.llm.provider import get_provider
from app.db.models import Analysis, Signal

logger = structlog.get_logger()


async def analyze_signal(
    signal: Signal,
    provider_name: str = "openai",
    model: str | None = None,
) -> Analysis:
    """Analyze a signal end-to-end.

    Steps:
    1. Extract key facts
    2. Classify sentiment
    3. Identify strategic themes
    4. Generate summary
    5. Calculate composite confidence
    6. Return AnalysisResult
    """
    logger.info(
        "Starting signal analysis",
        signal_id=signal.id,
        source_type=signal.source_type,
        content_length=len(signal.raw_content),
    )

    # Run analysis tasks (could be parallelized with asyncio.gather)
    facts_result = await extract_facts(signal.raw_content, provider_name, model)
    sentiment_result = await classify_sentiment(signal.raw_content, provider_name, model)
    themes_result = await identify_themes(signal.raw_content, provider_name, model)

    # Generate summary
    provider = get_provider(provider_name)
    summary_messages = build_summary_prompt(signal.raw_content, signal.title)
    summary_request = LLMRequest(messages=summary_messages, model=model, temperature=0.5)
    summary_response = await provider.complete(summary_request)
    summary = summary_response.content

    # Calculate composite confidence
    confidence = calculate_confidence(
        source_type=signal.source_type,
        content_length=len(signal.raw_content),
        facts=facts_result.facts,
        themes=themes_result.themes,
        llm_confidence=sentiment_result.confidence,
    )

    analysis = Analysis(
        id=str(uuid4()),
        signalId=signal.id,
        summary=summary,
        keyFacts=[f.model_dump() for f in facts_result.facts],
        sentiment=sentiment_result.sentiment.value,
        strategicThemes=[t.model_dump() for t in themes_result.themes],
        confidence=confidence,
        modelUsed=provider_name,
        analyzedAt=datetime.utcnow(),
    )

    logger.info(
        "Completed signal analysis",
        signal_id=signal.id,
        analysis_id=analysis.id,
        confidence=round(confidence, 3),
        sentiment=sentiment_result.sentiment,
        facts_count=len(facts_result.facts),
        themes_count=len(themes_result.themes),
    )

    return analysis

File: src/lib/ai/pipeline.ts
/**
 * Analysis pipeline orchestrator.
 * Translated from backend/app/analysis/pipeline.py
 */

import { logger } from "@/lib/logger";
import { extractFacts } from "./fact-extraction";
import { classifySentiment } from "./sentiment";
import { identifyThemes } from "./themes";
import { calculateConfidence } from "./confidence";
import { getProviderWithFailover } from "./provider";
import { buildSummaryPrompt } from "./prompts";
import { z } from "zod";
import type { ProviderName } from "./provider";
import type { SourceType, Sentiment } from "./types";

const SummaryResultSchema = z.object({
  summary: z.string(),
});

export interface AnalysisInput {
  id: string;
  sourceUrl: string;
  sourceType: SourceType;
  title: string;
  rawContent: string;
  publishedAt: Date | null;
  scrapedAt: Date;
  companyId: string;
  status: string;
  company?: {
    id: string;
    name: string;
    slug: string;
    ticker: string | null;
  };
}

export interface AnalysisResult {
  id: string;
  signalId: string;
  summary: string;
  keyFacts: Array<{
    text: string;
    category: string;
    confidence: number;
    sourceSentence?: string;
  }>;
  sentiment: Sentiment;
  strategicThemes: Array<{
    label: string;
    evidence: string[];
    correlationHints?: string[];
  }>;
  confidence: number;
  modelUsed: string;
  analyzedAt: Date;
}

/**
 * Run the full analysis pipeline on a signal.
 *
 * Steps:
 * 1. Extract facts from the signal text
 * 2. Classify sentiment
 * 3. Identify strategic themes
 * 4. Generate a summary
 * 5. Calculate composite confidence score
 * 6. Return structured analysis result
 */
export async function analyzeSignal(
  signal: AnalysisInput,
  providerName: ProviderName = "openai",
  model?: string
): Promise<AnalysisResult> {
  const startTime = Date.now();
  const log = logger.child({ signalId: signal.id, provider: providerName });

  log.info("analysis.pipeline.start", {
    sourceType: signal.sourceType,
    contentLength: signal.rawContent.length,
  });

  try {
    // Run fact extraction, sentiment, and themes in parallel
    const [factsResult, sentimentResult, themesResult] = await Promise.all([
      extractFacts(signal.rawContent, providerName, model),
      classifySentiment(signal.rawContent, providerName, model),
      identifyThemes(signal.rawContent, providerName, model),
    ]);

    log.debug("analysis.pipeline.parallel_complete", {
      facts_count: factsResult.facts.length,
      sentiment: sentimentResult.sentiment,
      themes_count: themesResult.themes.length,
    });

    // Generate summary
    const { provider } = getProviderWithFailover(providerName);
    const companyName = signal.company?.name ?? "the company";
    const summaryMessages = buildSummaryPrompt(signal.rawContent, companyName);
    const summaryResult = await provider.completeStructured(
      summaryMessages,
      SummaryResultSchema,
      { model, temperature: 0.4 }
    );

    // Calculate composite confidence
    const modelUsed = model ?? process.env.FAST_MODEL ?? "unknown";
    const confidence = calculateConfidence({
      sourceType: signal.sourceType,
      contentLength: signal.rawContent.length,
      facts: factsResult.facts,
      themes: themesResult.themes,
      llmConfidence: sentimentResult.confidence,
    });

    const latencyMs = Date.now() - startTime;

    log.info("analysis.pipeline.complete", {
      confidence: Math.round(confidence * 1000) / 1000,
      latency_ms: latencyMs,
    });

    return {
      id: crypto.randomUUID(),
      signalId: signal.id,
      summary: summaryResult.summary,
      keyFacts: factsResult.facts.map((f) => ({
        text: f.text,
        category: f.category,
        confidence: f.confidence,
        sourceSentence: f.source_sentence,
      })),
      sentiment: sentimentResult.sentiment,
      strategicThemes: themesResult.themes.map((t) => ({
        label: t.label,
        evidence: t.evidence,
        correlationHints: t.correlation_hints,
      })),
      confidence,
      modelUsed,
      analyzedAt: new Date(),
    };
  } catch (error) {
    log.error("analysis.pipeline.error", { error: String(error) });
    throw error;
  }
}
